lua/luasnip/extras/otf.lua

```diff
@@ -0,0 +1,86 @@
+local ls = require("luasnip")
+local cp = require("luasnip.util.functions").copy
+local p = require("luasnip.extras._parser_combinator")
+
+local T = { EOL = "EOL", TXT = "TXT", INP = "INP" }
+
+local chunk = p.any(
+	p.map(p.literal("$$"), function()
+		return { T.TXT, "$" }
+	end),
+	p.map(p.literal("\n"), function()
+		return { T.EOL }
+	end),
+	p.map(p.seq(p.literal("$"), p.pattern("%w+")), function(c)
+		return { T.INP, c[1] }
+	end),
+	p.map(p.pattern("[^\n$]*"), function(c)
+		return { T.TXT, c }
+	end)
+)
+
+local snippet_chunks = p.star(chunk)
+
+local function txt_to_snip(txt)
+	local t = ls.t
+	local sn = ls.s
+	local i = ls.i
+	local f = ls.f
+
+	-- The parser does not handle empty strings
+	if txt == "" then
+		return sn(t({ "" }))
+	end
+
+	local _, chunks, _ = snippet_chunks(txt, 1)
+
+	local current_text_arg = { "" }
+	local nodes = {}
+	local know_inputs = {}
+	local last_input_pos = 0
+
+	for _, part in ipairs(chunks) do
+		if part[1] == T.TXT then
+			current_text_arg[#current_text_arg] = current_text_arg[#current_text_arg]
+				.. part[2]
+		elseif #current_text_arg > 0 then
+			table.insert(nodes, t(current_text_arg))
+			current_text_arg = { "" }
+		end
+
+		if part[1] == T.EOL then
+			table.insert(current_text_arg, "")
+		elseif part[1] == T.INP then
+			local inp_pos = know_inputs[part[2]]
+			if inp_pos then
+				table.insert(nodes, f(cp, { inp_pos }))
+			else
+				last_input_pos = last_input_pos + 1
+				know_inputs[part[2]] = last_input_pos
+				table.insert(nodes, i(last_input_pos, part[2]))
+			end
+		end
+	end
+	if #current_text_arg > 1 or current_text_arg[0] ~= "" then
+		table.insert(nodes, t(current_text_arg))
+	end
+	return sn("", nodes)
+end
+
+local last_snip = nil
+local last_reg = nil
+
+local M = {}
+
+-- Create snippets On The Fly
+-- It's advaisable not to use the default register as luasnip will probably
+-- override it
+function M.otf()
```

Comment:
What do you think about passing the register used for generating the snippet to otf? That way one could potentially have multiple otf-snippets at a time.

Reply:
That's not needed `getreg()` uses the last specified register so you can do `"x` before calling it and that works as you want.

Reply:
Last used in general or last used named register? I think preserving the otf snippet over deletions or similar would be nice.

Reply:
It takes the value of `v:register` which is the last used register, that's why in the doc I'm specifying the register before calling the function in the shortcuts.

Reply:
Right, I'm thinking about expanding the recorded snippet multiple times, ie. I'd like to expand the recorded snippet 5 times, but after the third I delete a line, which makes another register v:register, so I'd lose the recorded snippet.
(At least that's how I understand the current implementation)

Reply:
Again, it will work fine if you use the shortcut as I said in the doc,  as that way it's always setting the correct register before calling the function.
```vim
nnoremap <leader>e  "p:lua require('luasnip.extras.otf').on_the_fly()<cr>
```
Note the  `"p`, you could use any other register there instead, like:
```vim
nnoremap <leader>e  "a:lua require('luasnip.extras.otf').on_the_fly()<cr>
```

Reply:
Ahh, I glanced over that and thought it was only for the initial yank/creation of the snippet, apologies :sweat_smile: